modules/social_features/social_email_broadcast/src/SocialEmailBroadcast.php

```diff
@@ -0,0 +1,110 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\social_email_broadcast;
+
+use Drupal\Core\Database\Connection;
+use Drupal\Core\Session\AccountInterface;
+
+/**
+ * Common service contains methods for "Social Email Broadcast".
+ */
+class SocialEmailBroadcast {
+
+  /**
+   * Immediate frequency.
+   */
+  const FREQUENCY_IMMEDIATELY = 'immediately';
+
+  /**
+   * None frequency.
+   */
+  const FREQUENCY_NONE = 'none';
+
+  /**
+   * Bulk emails user's frequencies table name.
+   */
+  const TABLE = 'user_email_send';
+
+  /**
+   * Constructs a SocialEmailBroadcast object.
+   */
+  public function __construct(
+    private readonly Connection $connection,
+  ) {}
+
+  /**
+   * Set bulk mailing user frequency.
+   *
+   * @param \Drupal\Core\Session\AccountInterface $account
+   *   The account entity.
+   * @param array $settings
+   *   The list of frequency settings.
+   *
+   * @throws \Exception
+   */
+  public function setBulkEmailUserSettings(AccountInterface $account, array $settings): void {
+    if (empty($settings)) {
+      return;
+    }
+
+    foreach ($settings as $name => $frequency) {
+      $query = $this->connection->merge(self::TABLE);
+      $query->fields([
+        'uid' => $account->id(),
+        'name' => $name,
+        'frequency' => $frequency,
+      ]);
+      $query->keys([
+        'uid' => $account->id(),
+        'name' => $name,
+      ]);
+      $query->execute();
+    }
+  }
+
+  /**
+   * Get all bulk mailing frequencies for given user.
+   *
+   * @param \Drupal\Core\Session\AccountInterface $account
+   *   The account entity.
+   *
+   * @return array
+   *   The list of frequencies associated by name.
+   *
+   * @throws \Exception
+   */
+  public function getBulkEmailUserSettings(AccountInterface $account): array {
+    return (array) $this->connection->select($alias = self::TABLE)
+      ->fields($alias, ['name', 'frequency'])
+      ->condition('uid', $account->id())
+      ->execute()
+      ?->fetchAllKeyed();
```

Comment:
- `execute()` can return NULL (see [docs](https://git.drupalcode.org/project/drupal/-/blob/11.x/core/lib/Drupal/Core/Database/Query/SelectInterface.php#L498))
- unfortunately I can't accept your suggestion because it makes run `$this->connection->select($alias = self::TABLE)...` twice 🙂

Reply:
> execute() can return NULL (see [docs](https://git.drupalcode.org/project/drupal/-/blob/11.x/core/lib/Drupal/Core/Database/Query/SelectInterface.php#L498))

Yes but this is only "if the query is not valid."

So from a business logic point of view, if the `execute` method returned `NULL` then that means there's a bug in our code. So we likely don't want to just provide an empty array but log some error instead.

Currently our code can be broken but it just returns an empty array without any complaints, so we likely find the bug years later because someone reports "Something is not happening when it should".

> unfortunately I can't accept your suggestion because it makes run $this->connection->select($alias = self::TABLE)... twice 🙂

This I don't understand? I replaced one line with another, so how would it suddenly make it run twice?